rust/src/common.rs

```diff
@@ -152,3 +152,152 @@ pub unsafe extern "C" fn rs_to_hex_sep(
     // overwrites last separator with final null char
     oslice[3 * islice.len() - 1] = 0;
 }
+
+use aho_corasick::AhoCorasick;
+
+#[derive(Debug,Clone)]
+struct AhoCorasickPatternData {
+    pat: Vec<u8>,
+    ci: bool,
+    offset: u16,
+    depth: u16,
+    sids: Vec<u32>,
+}
+
+impl AhoCorasickPatternData {
+    fn new(pat: Vec<u8>, ci: bool, sids: Vec<u32>, offset: u16, depth: u16) -> Self {
+        Self { pat, ci, sids, offset, depth }
+    }
+}
+
+#[derive(Default)]
+pub struct AhoCorasickStateBuilder {
+    /// vector of patterns. The final pattern id will depend on the position in this
+    /// vector, starting at 0.
+    patterns: Vec<Vec<u8>>,
+    /// Hash of patterns with their settings. Will be copied to AhoCorasickStateBuilder
+    /// in the prepare step.
+    pattern_data: Vec<AhoCorasickPatternData>,
+    /// track if we have case insensitive patterns. If so, we need to tell AC and
+    /// do a bit more work in validation.
+    has_ci: bool,
+}
+
+impl AhoCorasickStateBuilder {
+    fn new() -> Self {
+        Self { ..Default::default() }
+    }
+    fn add_pattern(&mut self, pat: Vec<u8>, ci: bool, sids: Vec<u32>, offset: u16, depth: u16) {
+        self.patterns.push(pat.clone());
+        if ci {
+            self.has_ci = true;
+        }
+        self.pattern_data.push(AhoCorasickPatternData::new(pat.clone(), ci, sids, offset, depth));
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn rs_mpm_acrs_new_builder() -> *mut std::os::raw::c_void {
```

Comment:
In addition to the unsafe notes from the previous PR...  We've been doing C style naming for functions that form the C API recently.. So `SCMpmAcRsNewBuilder` or whatever.

I guess this kind of breaks the way we've been looking at the impact of Rust functions tho.

Reply:
These functions would not need to be part of a public C API. C API should only expose it through the generic MPM support. So would that mean we drop the SC? `MpmAcRsNewBuilder`?

Reply:
If its no_mangle extern "C" its part of the public API, as far as linking goes which is why we want to name space under `SC`.

Reply:
Hmm can we have a way to make things not public from a libsuri perspective? This stuff is public not because I want it to be, but because the C-Rust FFI.

Reply:
Name, like `SCPrivate...` or `_SC` to make it clear its for private use only.  Additionally removing or `ifdef` out of the public headers, either would require some creative use or post-processing of the cbindgen output.  The namespacing is to avoid a collision, and the name scheme or ifdef is to avoid unintentional linking. Nothing we can do about intentional linking.